README.md
Medic
=====

[![Build Status](https://api.travis-ci.org/Rowno/medic.svg?branch=master)](https://travis-ci.org/Rowno/medic)
[![Dependency Status](https://david-dm.org/Rowno/medic/status.svg)](https://david-dm.org/Rowno/medic)

Perform bulk URL status checks and track changes.

![Medic screenshot](screenshot.png)


Getting Started
---------------

Install the Medic command line tool:
```bash
npm install -g medic
```

Create a `urls.txt` file:
```
---
cookies:
    - Location=nz
---

# Home pages

https://rolandwarmerdam.co.nz/
https://www.google.com/
https://twitter.com/
http://example.com/
```

Run medic:
```bash
medic urls.txt
```

Also supports piping:
```bash
cat urls.txt | medic
```


Usage
-----

### CLI

```
Usage: medic <path> [<options>]

Path:
    Path to a file containing a list of URLs. Each URL must be on a separate
    line and lines not starting with http:// or https:// are ignored.
    This parameter is ignored when URLs are being piped.
    File can also can also contain YAML front matter with the cookies option.

Options:
    -c  --concurrency Number of concurrent requests per origin. Defaults to 5.
    -h, --help        Show this help text.
    -o, --output      File path to output the results of this run to.
    -p, --compare     Path to a previous results file to compare against.
    -v, --version     Print medic's version.
```


### API

#### `.check(object options, [function callback])` -> `Promise`
Checks the status of all the passed URLs.

##### Options

###### `cookies`
Type: `array`

###### `urls`
Type: `array`

###### `onProgress`
Type: `function<object result>`

Gets called on each checked URL and is passed the result.

Example result:
```json
[
  {
    "url": "https://www.google.com/mail/",
    "statusCode": 200,
    "redirectUrl": "https://mail.google.com/mail/"
  }, {
    "url": "http://non-existent.example.com/",
    "error": "getaddrinfo ENOTFOUND"
  }
]
```


#### `.compare(object options)` -> `array`
Compares 2 result sets to find any changes.

Example return:
```json
[
  {
    "current": {
      "url": "https://reader.google.com/",
      "statusCode": 404
    },
    "previous": {
      "url": "https://reader.google.com/",
      "statusCode": 200
    }
  }
]
```

##### Options

###### `currentResults`
Type: `array`

###### `previousResults`
Type: `array`


Related
-------

 - [sitemap-urls][] - Extract URLs from an XML sitemap.


License
-------
Medic is released under the MIT license.

Copyright © 2015 Roland Warmerdam.



[sitemap-urls]: https://github.com/Rowno/sitemap-urls


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
This project follows [Semantic Versioning](http://semver.org).

## 3.1.0 / 2015-10-29
 - add: cookies option.
 - add: yaml front matter support to urls file (only for cookies option atm).
 - change: catch errors thrown in the cli.

## 3.0.0 / 2015-09-12
 - change: update dependencies.
 - add: node 4.0 support.
 - remove: node 0.10 support.

## 2.0.0 / 2015-03-04
 - add: --concurrency option.
 - change: set a default concurrency (maxSockets) to prevent DOSing sites on Node >=0.12.
 - change: --compare option shorthand to -p.
 - change: update dependencies.

## 1.4.0 / 2015-02-16
 - change: detect misc ASP.NET UnhandledException errors.

## 1.3.0 / 2015-01-25
 - change: output errors on stderr.
 - change: update dependencies.

## 1.2.0 / 2015-01-19
 - add: -v flag for printing the version number.
 - add: update-notifier.

## 1.1.0 / 2015-01-18
 - add: padding to the command line progress number.

## 1.0.0 / 2015-01-17
 - initial release.


test/cli.js
/* eslint-disable no-sync */
'use strict';
var path = require('path');
var fs = require('fs');
var exec = require('child_process').exec;
var expect = require('chai').expect;
var express = require('express');

var CLI = path.resolve(require('../package.json').bin.medic);
var URLS_FILE = path.join(__dirname, 'fixtures/urls.txt');
var TEMP_OUTPUT_FILE = path.join(__dirname, 'fixtures/temp-results.json');
var COMPARE_FILE = path.join(__dirname, 'fixtures/results-previous.json');
var PORT = 15000;
var COOKIE_STATUS = 400;
var HTML = '<!DOCTYPE html><html lang="en"><head><meta charset="UTF-8"><title>Test</title></head><body></body></html>';
var app = express();
var server;
var fixtureOutput = [
    '1/2  ✔  200  http://localhost:' + PORT + '/1/',
    '2/2  ✔  200  http://localhost:' + PORT + '/2/',
    '',
].join('\n');


app.use(function (req, res) {
    if (req.headers.cookie === 'Location=nz') {
        res.status(COOKIE_STATUS).send(HTML);
    } else {
        res.end(HTML);
    }
});


describe('cli', function () {
    before(function (done) {
        server = app.listen(PORT, 'localhost', done);
    });

    after(function () {
        if (fs.existsSync(TEMP_OUTPUT_FILE)) {
            fs.unlinkSync(TEMP_OUTPUT_FILE);
        }
        server.close();
    });


    it('should get status of urls in file', function (done) {
        var child = exec(
            CLI + ' ' + URLS_FILE,
            { cwd: __dirname },
            function (error, stdout, stderr) {
                if (error) {
                    return done(error);
                }

                expect(stdout, 'stdout').to.equal(fixtureOutput);
                expect(stderr, 'stderr').to.equal('');
                done();
            }
        );

        child.stdin.end();
    });


    it('should get status of urls passed through stdin', function (done) {
        var child = exec(
            CLI,
            { cwd: __dirname },
            function (error, stdout, stderr) {
                if (error) {
                    return done(error);
                }

                expect(stdout, 'stdout').to.equal(fixtureOutput);
                expect(stderr, 'stderr').to.equal('');
                done();
            }
        );

        fs.createReadStream(URLS_FILE, { encoding: 'utf8' }).pipe(child.stdin);
    });


    it('should output results to file', function (done) {
        var fixture = [{
            url: 'http://localhost:' + PORT + '/1/',
            statusCode: 200
        }, {
            url: 'http://localhost:' + PORT + '/2/',
            statusCode: 200
        }];

        var child = exec(
            CLI + ' ' + URLS_FILE + ' --output ' + TEMP_OUTPUT_FILE,
            { cwd: __dirname },
            function (error, stdout, stderr) {
                var outputFileContent;

                if (error) {
                    return done(error);
                }

                outputFileContent = fs.readFileSync(TEMP_OUTPUT_FILE, { encoding: 'utf8' });
                outputFileContent = JSON.parse(outputFileContent);

                expect(outputFileContent, 'outputted results').to.deep.equal(fixture);
                expect(stdout, 'stdout').to.equal(fixtureOutput);
                expect(stderr, 'stderr').to.equal('');
                done();
            }
        );

        child.stdin.end();
    });


    it('should log compare results', function (done) {
        var child;
        var fixture = [
            '',
            'Changes',
            '',
            '404 →  200  http://localhost:' + PORT + '/1/',
            '',
        ].join('\n');

        fixture = fixtureOutput + fixture;

        child = exec(
            CLI + ' ' + URLS_FILE + ' --compare ' + COMPARE_FILE,
            { cwd: __dirname },
            function (error, stdout, stderr) {
                if (error) {
                    return done(error);
                }

                expect(stdout, 'stdout').to.equal(fixture);
                expect(stderr, 'stderr').to.equal('');
                done();
            }
        );

        child.stdin.end();
    });


    it('should set cookies from yaml front matter', function (done) {
        var child;

        child = exec(
            CLI + ' ' + path.join(__dirname, 'fixtures/cookies.txt'),
            { cwd: __dirname },
            function (error, stdout, stderr) {
                if (error) {
                    return done(error);
                }

                expect(stdout, 'stdout').to.equal('1/1  ⚠  ' + COOKIE_STATUS + '  http://localhost:' + PORT + '/1/\n');
                expect(stderr, 'stderr').to.equal('');
                done();
            }
        );

        child.stdin.end();
    });
});


test/index.js
'use strict';
var expect = require('chai').expect;
var nock = require('nock');
var medic = require('../');

var SUCCESS = 200;
var MOVED = 301;
var NOT_FOUND = 404;
var HTML = '<!DOCTYPE html><html lang="en"><head><meta charset="UTF-8"><title>Test</title></head><body></body></html>';


describe('index', function () {
    describe('#check', function () {
        it('should get the status of urls', function () {
            var fixture = [{
                url: 'http://localhost/1/',
                statusCode: SUCCESS
            }, {
                url: 'http://localhost/2/',
                statusCode: NOT_FOUND
            }];

            nock('http://localhost')
                .get('/1/')
                .reply(SUCCESS, HTML)
                .get('/2/')
                .reply(NOT_FOUND, HTML);

            return medic.check({
                urls: [
                    'http://localhost/1/',
                    'http://localhost/2/'
                ]
            }).then(function (result) {
                expect(result).to.deep.equal(fixture);
            });
        });


        it('should track redirects', function () {
            var fixture = [{
                url: 'http://localhost/1/',
                statusCode: SUCCESS,
                redirectUrl: 'http://localhost/2/'
            }];

            nock('http://localhost')
                .get('/1/')
                .reply(MOVED, HTML, {
                    Location: 'http://localhost/2/'
                })
                .get('/2/')
                .reply(SUCCESS, HTML);

            return medic.check({
                urls: ['http://localhost/1/']
            }).then(function (result) {
                expect(result).to.deep.equal(fixture);
            });
        });


        it('should set status code to 500 for ASP.NET 500 error page', function () {
            var fixture = [{
                url: 'http://localhost/errors/500.aspx?aspxerrorpath=/1/',
                statusCode: 500
            }];

            nock('http://localhost')
                .get('/errors/500.aspx?aspxerrorpath=/1/')
                .reply(SUCCESS, HTML);

            return medic.check({
                urls: [
                    'http://localhost/errors/500.aspx?aspxerrorpath=/1/'
                ]
            }).then(function (result) {
                expect(result).to.deep.equal(fixture);
            });
        });


        it('should call onProgress function with each URL check', function () {
            var fixture = [{
                url: 'http://localhost/1/',
                statusCode: SUCCESS
            }, {
                url: 'http://localhost/2/',
                statusCode: NOT_FOUND
            }];
            var count = 0;

            nock('http://localhost')
                .get('/1/')
                .reply(SUCCESS, HTML)
                .get('/2/')
                .reply(NOT_FOUND, HTML);

            return medic.check({
                urls: [
                    'http://localhost/1/',
                    'http://localhost/2/'
                ],
                onProgress: function (result) {
                    expect(result, 'progress result').to.deep.equal(fixture[count]);
                    count += 1;
                }
            }).then(function (result) {
                expect(count, 'progress count').to.equal(2);
                expect(result, 'final result').to.deep.equal(fixture);
            });
        });


        it('should support standard callbacks', function (done) {
            var fixture = [{
                url: 'http://localhost/1/',
                statusCode: SUCCESS
            }];

            nock('http://localhost')
                .get('/1/')
                .reply(SUCCESS, HTML);

            medic.check({
                urls: ['http://localhost/1/']
            }, function (error, result) {
                if (error) {
                    return done(error);
                }

                expect(result, 'final result').to.deep.equal(fixture);
                done();
            });
        });


        it('should set cookies', function (done) {
            var fixture = [{
                url: 'http://localhost/1/',
                statusCode: SUCCESS
            }];

            nock('http://localhost', { reqheaders: { Cookie: 'Location=nz' } })
                .get('/1/')
                .reply(SUCCESS, HTML);

            medic.check({
                cookies: ['Location=nz'],
                urls: ['http://localhost/1/'],
            }, function (error, result) {
                if (error) {
                    return done(error);
                }

                expect(result, 'final result').to.deep.equal(fixture);
                done();
            });
        });
    });


    describe('#compare', function () {
        it('should return results that have different status codes', function () {
            var fixtureCurrent = [{
                url: 'http://localhost/1/',
                statusCode: SUCCESS
            }, {
                url: 'http://localhost/2/',
                statusCode: NOT_FOUND
            }];
            var fixturePrevious = [{
                url: 'http://localhost/1/',
                statusCode: SUCCESS
            }, {
                url: 'http://localhost/2/',
                statusCode: SUCCESS
            }];
            var fixtureCompare = [{
                current: {
                    url: 'http://localhost/2/',
                    statusCode: NOT_FOUND
                },
                previous: {
                    url: 'http://localhost/2/',
                    statusCode: SUCCESS
                }
            }];

            var result = medic.compare({
                currentResults: fixtureCurrent,
                previousResults: fixturePrevious
            });

            expect(result).to.deep.equal(fixtureCompare);
        });
    });
});


lib/index.js
'use strict';
var _ = require('lodash');
var bluebird = require('bluebird');
var request = require('request');
bluebird.promisifyAll(request);


/**
 * Checks the status of all the passed URLs.
 * @param  {object}   opt
 * @param  {function} callback
 * @return {Promise}
 */
function check(opt, callback) {
    var urlRequests;

    var options = _.defaults(opt, {
        urls: [],
        cookies: [],
        onProgress: _.noop,
    });

    urlRequests = options.urls.map(function checkUrl(requestUrl) {
        var jar = request.jar();

        options.cookies.forEach(function (cookie) {
            jar.setCookie(request.cookie(cookie), requestUrl);
        });

        return request.getAsync({
            url: requestUrl,
            jar: jar, // Store cookies (separate for each url)
            gzip: true,
            headers: { // Pretend to be Chrome for any bad user agent sniffers
                'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_3) AppleWebKit/537.31 (KHTML, like Gecko) Chrome/26.0.1410.65 Safari/537.31'
            }
        }).spread(function handleResponse(response, body) {
            var responseUrl = response.request.uri.href;
            var result = {
                url: requestUrl,
                statusCode: response.statusCode
            };

            // Hack for incorrent status code on ASP.NET 500 error page and
            // misc ASP.NET server-side errors that bypass the 500 error page
            if (responseUrl.indexOf('500.aspx') >= 0 ||
                body.indexOf('UnhandledException') >= 0) {
                result.statusCode = 500;
            }

            if (requestUrl !== responseUrl) {
                result.redirectUrl = responseUrl;
            }

            return result;
        }).error(function handleError(error) {
            return {
                url: requestUrl,
                error: error.cause.message
            };
        }).tap(options.onProgress);
    });

    return bluebird.all(urlRequests).nodeify(callback);
}


/**
 * Compares 2 result sets to find any changes.
 * @param  {object} opt
 * @return {array}
 */
function compare(opt) {
    var currentResults;
    var previousResults;
    var changedResults = [];

    var options = _.defaults(opt, {
        currentResults: [],
        previousResults: []
    });

    currentResults = options.currentResults;
    previousResults = options.previousResults;

    currentResults.forEach(function compareResult(currentResult) {
        var previousResult = _.find(previousResults, function findPreviousResult(result) {
            return currentResult.url === result.url;
        });

        if (!previousResult) {
            return;
        }

        if (currentResult.statusCode !== previousResult.statusCode) {
            changedResults.push({
                current: currentResult,
                previous: previousResult
            });
        }
    });

    return changedResults;
}


exports.check = check;
exports.compare = compare;


lib/cli.js
#!/usr/bin/env node

/* eslint-disable no-sync, no-process-exit */
'use strict';
var fs = require('fs');
var http = require('http');
var https = require('https');
var path = require('path');

var _ = require('lodash');
var chalk = require('chalk');
var figures = require('figures');
var frontMatter = require('front-matter');
var meow = require('meow');
var stdin = require('get-stdin');
var updateNotifier = require('update-notifier');

var medic = require('../');
var pkg = require('../package.json');

var NOT_FOUND = 200;
var INTERNAL_ERROR = 500;
var HELP_FILE_PATH = path.join(__dirname, 'help.txt');
var cli;


updateNotifier({ pkg: pkg }).notify();

cli = meow({
    pkg: pkg,
    help: fs.readFileSync(HELP_FILE_PATH, { encoding: 'utf8' }).trim()
}, {
    alias: {
        compare: 'p',
        concurrency: 'c',
        help: 'h',
        output: 'o',
        version: 'v',
    },
    default: {
        concurrency: 5
    }
});

// Always set a concurrency to avoid DOSing sites on Node >=0.12
http.globalAgent.maxSockets = cli.flags.concurrency;
https.globalAgent.maxSockets = cli.flags.concurrency;


/**
 * Add the terminal status color for the given result to a string.
 * @param {number}  result
 * @param {string}  message
 * @return {string}
 */
function addStatusColor(result, message) {
    var coloredMessage = message;

    if (result.statusCode === NOT_FOUND) {
        coloredMessage = chalk.green(message);
    } else if (result.error || result.statusCode === INTERNAL_ERROR) {
        coloredMessage = chalk.red(message);
    } else {
        coloredMessage = chalk.yellow(message);
    }

    return coloredMessage;
}


stdin().then(function (stdinUrls) {
    var compareFileContent;
    var compareFilePath;
    var frontMatterData;
    var outputFilePath;
    var previousResults = [];
    var progress = 1;
    var urls = [];
    var urlsFilePath;
    var urlsNumLength;
    var urlsString;


    // Require stdin or a urls file
    if (!stdinUrls && !_.isString(cli.input[0])) {
        cli.showHelp();
        process.exit(1);
    }

    // Try reading urls file if no stdin
    if (stdinUrls) {
        urlsString = stdinUrls;
    } else {
        urlsFilePath = path.resolve(cli.input[0]);
        if (!fs.existsSync(urlsFilePath) || !fs.statSync(urlsFilePath).isFile()) {
            console.error('File doesn\'t exist:', urlsFilePath);
            process.exit(1);
        }

        urlsString = fs.readFileSync(urlsFilePath, { encoding: 'utf8' });
    }


    frontMatterData = frontMatter(urlsString);


    // Parse urls from string
    urls = frontMatterData.body.match(/^https?:\/\/(.*?)$/mg) || [];
    urlsNumLength = String(urls.length).length;


    if (_.isString(cli.flags.compare)) {
        // Check compare file exists
        compareFilePath = path.resolve(cli.flags.compare);
        if (!fs.existsSync(compareFilePath) || !fs.statSync(compareFilePath).isFile()) {
            console.error('File doesn\'t exist:', compareFilePath);
            process.exit(1);
        }

        // Parse results from file
        compareFileContent = fs.readFileSync(compareFilePath, { encoding: 'utf8' });
        previousResults = JSON.parse(compareFileContent);
    }


    if (_.isString(cli.flags.output)) {
        outputFilePath = path.resolve(cli.flags.output);
    }


    /**
     * Logs the check progress to the terminal.
     * @param {object} result
     */
    function progressLog(result) {
        var icon;
        var progressNumLength = String(progress).length;
        var padding = _.range(0, urlsNumLength - progressNumLength, 0).join('');

        if (result.statusCode === NOT_FOUND) {
            icon = figures.tick;
        } else if (result.error || result.statusCode === INTERNAL_ERROR) {
            icon = figures.cross;
        } else {
            icon = figures.warning;
        }

        console.log(addStatusColor(result,
            padding + progress + '/' + urls.length +
            '  ' +
            icon +
            '  ' +
            (result.error ? 'err' : result.statusCode) +
            '  ' +
            result.url
        ));

        progress += 1;
    }

    return medic.check({
        cookies: frontMatterData.attributes.cookies,
        urls: urls,
        onProgress: progressLog
    }).then(function processResults(results) {
        var resultsString;
        var compareResults = [];

        // Do compare if compare file was given
        if (previousResults) {
            compareResults = medic.compare({
                currentResults: results,
                previousResults: previousResults
            });
        }

        // Ouput changes if there were any
        if (compareResults.length > 0) {
            console.log();
            console.log(chalk.bold('Changes'));
            console.log();

            compareResults.forEach(function logCompareResult(result) {
                var previousStatusCode = result.previous.error ? 'err' : result.previous.statusCode;
                var currentStatusCode = result.current.error ? 'err' : result.current.statusCode;

                console.log(
                    addStatusColor(result.previous, previousStatusCode),
                    figures.arrowRight,
                    '',
                    addStatusColor(result.current, currentStatusCode),
                    '',
                    result.previous.url
                );
            });
        }

        // Output results if output path was given
        if (outputFilePath) {
            resultsString = JSON.stringify(results, null, 2);
            fs.writeFileSync(outputFilePath, resultsString);
        }
    });
}).catch(function (error) {
    console.error(error.stack || error);
    process.exit(1);
});
